[FEATURE] Chat field

burtenshaw opened the pull request:
This is feature PR for the chat field.

It contains changes for:
- frontend
- sdk
- server

We need to work on:
- docs

jfcalvo committed:
feat: add support to new fields of type image

jfcalvo committed:
chore: update CHANGELOG.md

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

jfcalvo committed:
improvement: validate that records values with associated image fields has correct URLs

jfcalvo committed:
chore: update CHANGELOG.md

jfcalvo committed:
chore: improve import

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

frascuchon committed:
[FEATURE] `argilla server`: Add support to image fields on search engine (#5355)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR adds support to indexing datasets and records with image fields.

Note: A datasets reindex is required with this change.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

jfcalvo committed:
feat: add length validation for record image field values

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

frascuchon committed:
feat: Add new ImageFieldSettings

frascuchon committed:
feat: Define new ImageField and a Abstract Field class

frascuchon committed:
refactor: Using generic Field instead of TextField

frascuchon committed:
tests: Add some basic unit tests

frascuchon committed:
tests: Add some image fields to tests

frascuchon committed:
ci: Using proper server version

frascuchon committed:
chore: remove type alias

jfcalvo committed:
feat: add MIME type validation for records using image fields values with Data URLs

jfcalvo committed:
Merge branch 'feat/add-image-field-support' into feat/argilla/add-support-to-create-image-fields

frascuchon committed:
[FEATURE] `argilla`: add support to create image fields (#5372)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR adds support to work with image fields using the SDK. 

## TODO (as separate PRs)

- [ ] Add documentation section
- [ ] Verify behavior using from/to_hub methods
- [ ] Add tooling to read data URLs from image files/folders (based on
logic defined in `image_to_html`)
- [ ] Add some media type validation (supported image types) 
- [ ] [Optional] Add tooling to re-scale images. The lower the image
sizes, the better the UI work. We can define a transparent re-scaling
process to support large images (currently the limit size in the backend
is 5MB for data URL)


**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: José Francisco Calvo <jose@argilla.io>

burtenshaw committed:
Update argilla-server/src/argilla_server/search_engine/commons.py

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

burtenshaw committed:
Merge branch 'develop' into feat/add-image-field-support

burtenshaw committed:
chore: update changelog

burtenshaw committed:
fix: merged error self >> cls

burtenshaw committed:
fix: merged error self >> cls

burtenshaw committed:
Merge branch 'feat/add-image-field-support' of https://github.com/argilla-io/argilla into feat/add-image-field-support

burtenshaw committed:
fix: merge mistake - suggestions at self not record

burtenshaw committed:
fix: merge mistake: more uncaught refactoring to class methods

damianpumar committed:
🌅 Image field (#5389)

This PR close this issue:
https://github.com/argilla-io/argilla/issues/5278

---------

Co-authored-by: Leire Aguirre <leire@argilla.io>

burtenshaw committed:
[FEATURE] [SERVER] chat field support server (#5377)

# Description
This PR adds support for a chat field to the server.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->


- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

Tests are updated

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

burtenshaw committed:
Merge branch 'feat/add-image-field-support' into feat/chat-field-support

burtenshaw committed:
[FEATURE] [SDK] chat field support sdk (#5378)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR implements chat field support on the SDK.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

damianpumar committed:
💬 Chat field frontend (#5388)

- [x] Markdown support
- [x] LaTex Support
- [x] Multiple agents
- [x] Custom colors for bubble chat
- [ ] Same color for single label (TBD)  
<img width="899" alt="Screenshot 2024-08-08 at 18 05 24"
src="https://github.com/user-attachments/assets/428a0f17-9860-4cd9-9eab-6e1edef404fc">

jfcalvo commented on argilla-frontend/v1/domain/entities/field/Field.ts:
```diff
@@ -14,12 +14,17 @@ export class Field {
     public settings: any
   ) {
     this.initializeOriginal();
+    console.log(this);
```
```suggestion
```

jfcalvo commented on argilla-server/CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ These are the section headers that we use:
 ### Added
 
 - Added new `image` type dataset field supporting URLs and Data URLs. ([#5279](https://github.com/argilla-io/argilla/pull/5279))
+- Added new `chat` type dataset field supporting chat-like conversations. ([#5375](https://github.com/argilla-io/argilla/pull/5375))
```
```suggestion
- Added new `chat` type dataset field supporting chat-like conversations. ([#5376](https://github.com/argilla-io/argilla/pull/5376))
```

jfcalvo commented on argilla-server/src/argilla_server/api/schemas/v1/records.py:
```diff
@@ -82,7 +82,7 @@ class Config:
 
 class RecordCreate(BaseModel):
-    fields: Dict[str, Union[StrictStr, None]]
+    fields: Dict[str, Union[StrictStr, None, List[Dict[StrictStr, StrictStr]]]]
```
I believe we should define a Pydantic model specific for the chat field.

So we define the `fields` attribute with something similar to:

```suggestion
    fields: Dict[str, Union[StrictStr, List[RecordChatField], None]]
```

And with `RecordChatField` being:

```python
class RecordChatField(BaseModel):
    role: str
    content: str
```

You can add validations to `RecordChatField`, define a `RecordChatFieldCreate` and `RecordChatFieldUpdate` if necessary, etc.

If necessary an alternative to this is define a `RecordChatFieldValue` as a list of `RecordChatField`'s with validations:

```suggestion
    fields: Dict[str, Union[StrictStr, RecordChatFieldValue, None]]
```

And with `RecordChatFieldValue` being:

```python
RecordChatFieldValue = Annotated[conlist(item_type=RecordChatField, min_items=1), Field(..., description="...")] 
```

burtenshaw replied:
I agree that a Pydantic model would be ideal here. 

However, we want to allow the object to have more keys/values than just 'role' and 'content' because these structures aren't fixed. Let's discuss this.

jfcalvo commented on argilla-server/src/argilla_server/validators/records.py:
```diff
@@ -126,6 +128,38 @@ def _validate_image_field(cls, field_name: str, field_value: Union[str, None]) -
+        if len(field_value) > CHAT_FIELD_MAX_LENGTH:
+            raise UnprocessableEntityError(
+                f"chat field {field_name!r} value is exceeding the maximum length of {CHAT_FIELD_MAX_LENGTH} characters"
+            )
```
I don't get this. Is not the chat field value a list of objects? Then should the message be changed?

burtenshaw replied:
Yes. This message is wrong. It should be 'max messages'. I will fix when we resolve the pydantic decision.

jfcalvo commented on argilla-server/src/argilla_server/validators/records.py:
```diff
@@ -126,6 +128,38 @@ def _validate_image_field(cls, field_name: str, field_value: Union[str, None]) -
+            cls._validate_chat_field(field.name, fields.get(field.name))
+
+    @classmethod
+    def _validate_chat_field(cls, field_name: str, field_value: Union[str, None]) -> None:
```
Maybe we don't need some of these validations if you add a Pydantic model.

burtenshaw committed:
Update argilla-server/CHANGELOG.md

Co-authored-by: José Francisco Calvo <jose@argilla.io>

burtenshaw committed:
fix: remove logging in Field

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'develop' into feat/chat-field-support

frascuchon committed:
Fix merge problems

leiyre committed:
fix plugin removed after merge

leiyre committed:
[UI] update UI for chat field (#5473)

frascuchon committed:
Merge branch 'develop' into feat/chat-field-support

leiyre committed:
fix renamed component

leiyre committed:
fix missing type

leiyre committed:
show use_markdown setting when field is chat type

frascuchon committed:
fix: revert test changes

leiyre committed:
differentiated color for roles with similar names

leiyre committed:
Merge branch 'feat/chat-field-support' of https://github.com/argilla-io/argilla into feat/chat-field-support

leiyre committed:
ui qa

leiyre committed:
ui color qa and fix text overflow

frascuchon committed:
[ENHANCEMENT] add markdown support for chat fields (#5482)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR adds changes in SDK and server to support markdown configuration
for `ChatField` by using the `use_markdown` flag in the same way as
`TextField`

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: Damián Pumar <damianpumar@gmail.com>
Co-authored-by: Leire Aguirre <leire@argilla.io>
Co-authored-by: José Francisco Calvo <jose@argilla.io>

leiyre committed:
ui remove title margin

leiyre committed:
Merge branch 'feat/chat-field-support' of https://github.com/argilla-io/argilla into feat/chat-field-support

frascuchon committed:
Merge branch 'develop' into feat/chat-field-support

leiyre committed:
update chat field styling to alternate roles on the chat items

leiyre committed:
add highlighting on search

burtenshaw committed:
[ENHANCEMENT] Use pydantic for chat field in server and SDK (#5450)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR use pydantic in the server and sdk to validate schema of chat
field values.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: José Francisco Calvo <josefranciscocalvo@gmail.com>

leiyre committed:
ui update field title color

leiyre committed:
Merge branch 'feat/chat-field-support' of https://github.com/argilla-io/argilla into feat/chat-field-support

leiyre committed:
ui update colors

leiyre committed:
ui qa

frascuchon committed:
Merge branch 'develop' into feat/chat-field-support
